File: 11723/cpp/boj_11723.cpp
// 집합
// https://www.acmicpc.net/problem/11723

#include <iostream>
#include <cstdint>
#include <string>
#include <bitset>


void solution()
{
    using namespace std;
    bitset<20> set;

    uint32_t m;
    cin >> m;

    for (uint32_t i = 0; i < m; ++i)
    {
        string command;
        cin >> command;

        if (command == "add")
        {
            uint32_t x;
            cin >> x;
            set.set(x - 1);
        }
        else if (command == "remove")
        {
            uint32_t x;
            cin >> x;
            set.reset(x - 1);
        }
        else if (command == "check")
        {
            uint32_t x;
            cin >> x;
            cout << set.test(x - 1) << '\n';
        }
        else if (command == "toggle")
        {
            uint32_t x;
            cin >> x;
            set.flip(x - 1);
        }
        else if (command == "all")
        {
            set.set();
        }
        else if (command == "empty")
        {
            set.reset();
        }
    }
}

int main()
{
    std::cin.tie(nullptr);
    std::cout.tie(nullptr);
    std::ios::sync_with_stdio(false);

    solution();

    return 0;
}

File: 11723/python/boj_11723.py
# 집합
# https://www.acmicpc.net/problem/11723


# 메모리 초과되서 안됨
from typing import Callable


class BitSet:
    def __init__(self):
        self.value: int = 0

    @staticmethod
    def to_bit(in_value: str) -> int:
        return 1 << int(in_value) - 1

    def add(self, in_value: str):
        self.value |= BitSet.to_bit(in_value)

    def remove(self, in_value: str):
        self.value &= ~BitSet.to_bit(in_value)

    def check(self, in_value: str):
        print(int((self.value & BitSet.to_bit(in_value)) != 0))

    def toggle(self, in_value: str):
        self.value ^= BitSet.to_bit(in_value)

    def all(self):
        self.value = (1 << 20) - 1

    def empty(self):
        self.value = 0


def solution(it_next: Callable[[], str]):
    my_set = BitSet()

    n = int(it_next())
    for _ in range(n):
        cmd, *args = it_next().split()
        getattr(my_set, cmd)(*args)


if __name__ == "__main__":
    solution(iter(open(0).read().splitlines()).__next__)
